src/tezos_interop/run_entrypoint.ml

```diff
@@ -0,0 +1,218 @@
+open Helpers
+open Crypto
+open Tezos
+
+type id = int [@@deriving yojson]
+let initial_id = 0
+module Id_map = Map.Make (Int)
+
+module Request = struct
+  type transaction = {
+    rpc_node : string;
+    secret : string;
+    confirmation : int;
+    destination : string;
+    entrypoint : string;
+    payload : Yojson.Safe.t;
+  }
+
+  type t = {
+    id : id;
+    content : transaction;
+  }
+
+  let to_yojson request =
+    let module T = struct
+      type t = {
+        id : id;
+        rpc_node : string;
+        secret : string;
+        confirmation : int;
+        destination : string;
+        entrypoint : string;
+        payload : Yojson.Safe.t;
+      }
+      [@@deriving to_yojson]
+    end in
+    let { id; content } = request in
+    let { rpc_node; secret; confirmation; destination; entrypoint; payload } =
+      content in
+    T.to_yojson
+      { id; rpc_node; secret; confirmation; destination; entrypoint; payload }
+end
+
+module Response = struct
+  type transaction =
+    | Applied     of { hash : string }
+    | Failed      of { hash : string }
+    | Skipped     of { hash : string }
+    | Backtracked of { hash : string }
+    | Unknown     of { hash : string }
+    | Error       of string
+  type t = {
+    id : id;
+    content : transaction;
+  }
+
+  let transaction_of_yojson json =
+    let module T = struct
+      type t = { status : string }
+      and finished = { hash : string }
+      and error = { error : string } [@@deriving of_yojson { strict = false }]
+    end in
+    let finished make =
+      let%ok { hash } = T.finished_of_yojson json in
+      Ok (make hash) in
+    let%ok { status } = T.of_yojson json in
+    match status with
+    | "applied" -> finished (fun hash -> Applied { hash })
+    | "failed" -> finished (fun hash -> Failed { hash })
+    | "skipped" -> finished (fun hash -> Skipped { hash })
+    | "backtracked" -> finished (fun hash -> Backtracked { hash })
+    | "unknown" -> finished (fun hash -> Unknown { hash })
+    | "error" ->
+      let%ok { error } = T.error_of_yojson json in
+      Ok (Error error)
+    | _ -> Error "invalid status"
+
+  let of_yojson json =
+    let module T = struct
+      type header = { id : int } [@@deriving of_yojson { strict = false }]
+    end in
+    let%ok { id } = T.header_of_yojson json in
+    let%ok content = transaction_of_yojson json in
+    Ok { id; content }
+end
+
+module Process : sig
+  type t
```

Comment:
What is the purpose of defining the sig both here and in the mli?

Reply:
Why not? It's an abstraction layer ensuring that inside of run_entrypoint.ml process is not used in the wrong way. And if it's here it must be in the mli.